tracepoint_decode/src/perf_abi.rs

```diff
@@ -0,0 +1,1409 @@
+
+    /// exclude_callchain_kernel: exclude kernel callchains
+    pub const ExcludeCallchainKernel: Self = Self(1 << 21);
+
+    /// exclude_callchain_user: exclude user callchains
+    pub const ExcludeCallchainUser: Self = Self(1 << 22);
+
+    /// mmap2: include mmap with inode data
+    pub const Mmap2: Self = Self(1 << 23);
+
+    /// comm_exec: flag comm events that are due to an exec
+    pub const CommExec: Self = Self(1 << 24);
+
+    /// use_clockid: use @clockid for time fields
+    pub const UseClockId: Self = Self(1 << 25);
+
+    /// context_switch: context switch data
+    pub const ContextSwitch: Self = Self(1 << 26);
+
+    /// write_backward: Write ring buffer from end to beginning
+    pub const WriteBackward: Self = Self(1 << 27);
+
+    /// namespaces: include namespaces data
+    pub const Namespaces: Self = Self(1 << 28);
+
+    /// ksymbol: include ksymbol events
+    pub const Ksymbol: Self = Self(1 << 29);
+
+    /// bpf_event: include bpf events
+    pub const BpfEvent: Self = Self(1 << 30);
+
+    /// aux_output: generate AUX records instead of events
+    pub const AuxOutput: Self = Self(1 << 31);
+
+    /// cgroup: include cgroup events
+    pub const Cgroup: Self = Self(1 << 32);
+
+    /// text_poke: include text poke events
+    pub const TextPoke: Self = Self(1 << 33);
+
+    /// build_id: use build id in mmap2 events
+    pub const BuildId: Self = Self(1 << 34);
+
+    /// inherit_thread: children only inherit if cloned with CLONE_THREAD
+    pub const InheritThread: Self = Self(1 << 35);
+
+    /// remove_on_exec: event is removed from task on exec
+    pub const RemoveOnExec: Self = Self(1 << 36);
+
+    /// sigtrap: send synchronous SIGTRAP on event
+    pub const Sigtrap: Self = Self(1 << 37);
+
+    /// Returns true if (self & mask) != 0.
+    pub const fn has_flag(self, mask: Self) -> bool {
+        0 != (self.0 & mask.0)
+    }
+
+    /// Returns `self | other`.
+    pub const fn or(self, other: Self) -> Self {
+        Self(self.0 | other.0)
+    }
+}
+
+impl From<u64> for PerfEventAttrOptions {
+    fn from(val: u64) -> Self {
+        Self(val)
+    }
+}
+
+impl From<PerfEventAttrOptions> for u64 {
+    fn from(val: PerfEventAttrOptions) -> Self {
+        val.0
+    }
+}
+
+/// perf_event_attr: Event's collection parameters.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default)]
+pub struct PerfEventAttr {
+    /// type:
+    /// Major type: hardware/software/tracepoint/etc.
+    pub attr_type: PerfEventAttrType,
+
+    /// size:
+    /// Size of the attr structure, for fwd/bwd compat.
+    pub size: PerfEventAttrSize,
+
+    /// config:
+    /// Type-specific configuration information.
+    pub config: u64,
+
+    /// sample_time: union of sample_period and sample_freq.
+    pub sample_time: u64,
+
+    /// sample_type
+    pub sample_type: PerfEventAttrSampleType,
+
+    /// read_format
+    pub read_format: PerfEventAttrReadFormat,
+
+    /// In C, this is a bit-field of various options:
```

Comment:
Not sure this is a useful comment.

Reply:
This is an ABI struct, and my intent is to help people map this back to the fields in the original perf_event_attr struct in perf_event.h. If somebody is searching "where does the `PerfEventAttr` struct define the `exclude_idle` flag" they will hit this comment and see that it is in the `options` field.

https://github.com/torvalds/linux/blob/5189dafa4cf950e675f02ee04b577dfbbad0d9b1/include/uapi/linux/perf_event.h#L414